source/levelcelview.cpp

```diff
@@ -178,6 +217,92 @@ void LevelCelView::framePixelClicked(quint16 x, quint16 y)
     }
 }
 
+void LevelCelView::insertFrames(QStringList filePaths, bool append)
+{
+    int prevFrameCount = this->gfx->getFrameCount();
+
+    if (append) {
+        // append the frame(s)
+        for (int i = 0; i < filePaths.count(); i++) {
+            int index = this->gfx->getFrameCount();
+            D1GfxFrame *frame = this->gfx->insertFrame(index, filePaths[i]);
+            if (frame != nullptr) {
+                frame->setFrameType(D1CEL_FRAME_TYPE::TransparentSquare);
+            }
+        }
+    } else {
+        // insert the frame(s)
+        for (int i = 1; i <= filePaths.count(); i++) {
+            D1GfxFrame *frame = this->gfx->insertFrame(this->currentFrameIndex, filePaths[filePaths.count() - i]);
+            if (frame != nullptr) {
+                frame->setFrameType(D1CEL_FRAME_TYPE::TransparentSquare);
+            }
+        }
+        // shift references + add default frame type
+        int deltaFrameCount = this->gfx->getFrameCount() - prevFrameCount;
+        if (deltaFrameCount > 0) {
+            unsigned refIndex = this->currentFrameIndex + 1;
+            // shift frame indices of the subtiles
+            for (int i = 0; i < this->min->getSubtileCount(); i++) {
+                QList<quint16> &frameIndices = this->min->getCelFrameIndices(i);
+                for (int n = 0; n < frameIndices.count(); n++) {
+                    if (frameIndices[n] >= refIndex) {
+                        frameIndices[n] += deltaFrameCount;
+                    }
+                }
+            }
+        }
+    }
+    // update the view
+    this->initialize(this->gfx, this->min, this->til, this->sol, this->amp);
+    this->displayFrame();
+}
+
+void LevelCelView::removeCurrentFrame()
+{
+    // check if the frame is used
+    int user = -1;
+    unsigned refIndex = this->currentFrameIndex + 1;
+    for (int i = 0; i < this->min->getSubtileCount() && user < 0; i++) {
+        QList<quint16> &frameIndices = this->min->getCelFrameIndices(i);
+        for (quint16 frameIdx : frameIndices) {
+            if (frameIdx == refIndex) {
+                user = i;
+                break;
+            }
+        }
+    }
+    if (user >= 0) {
+        QMessageBox::StandardButton reply;
+        reply = QMessageBox::question(nullptr, "Confirmation", "The frame is used by Subtile " + QString::number(user + 1) + " (and maybe others). Are you sure you want to proceed?", QMessageBox::Yes | QMessageBox::No);
```

Comment:
Should removing it also remove the subtiles, does the user need to manually correct the affected subtiles for thing to be valid again?

Reply:
The subtile is still valid. The frame is replaced with a 'transparent' (0) micro. I think a subtile removal should be an another function.